global-script.lua

```diff
@@ -3311,3 +3464,284 @@ function tCompare(t1,t2)
+            pattern = color .. "'s (.*)",
+            bagContents = {},
+            defendBag = defendBags[oppositeColor],
+            oppositeColor = oppositeColor
+        }
+    end
+    local colors = {
+        from = initData(fromColor, 1, toColor),
+        to = initData(toColor, 2, fromColor)
+    }
+
+    -- If both bags are full, there's not a lot of work to do.
+    -- Unfortunately, we still need to loop through other things because of defend tokens that aren't in bags.
+    local fastSwap = (colors.from.qty == 20 and colors.to.qty == 20)
+    -- Just bail out fast.
+
+    selectedColors[fromColor], selectedColors[toColor] = selectedColors[toColor], selectedColors[fromColor]
+
+    if not fastSwap then
+        -- Remove any items still in the bags
+        -- NOTE: TTS's documentation suggests we may need to wait a physics frame after doing this,
+        -- but this seems to work fine without doing that.  If something goes awry with this code in the future,
+        -- you might try re-adding the delay.
+        for color,data in pairs(colors) do
+            for i = 1,data.qty do 
+                local obj = data.bag.takeObject({
+                    sound=false,
+                    position={x=data.ix*2, z=200, y=i*2}    -- Chosen to be out-of-the-way and to prevent items from stacking.
+                })
+                table.insert(data.bagContents, obj)
+            end
+        end
+    end
+
+    -- Pass 1: Iterate over all objects looking for "<color>'s X".
+    -- Make a note of what we find and what tint it is.  Handle Defence tokens in this pass.
+    local match = string.match  -- Performance
+    local name, suffix
+    for _,obj in pairs(getAllObjects()) do
+        name = obj.getName()
+        if name then
+            for _,data in pairs(colors) do 
+                suffix = match(name, data.pattern)
+                if suffix then
+                    if suffix == 'Defence' then  -- Defense tokens are special.
+                        local state = obj.getStateId()
+                        local attrs = {position = obj.getPosition(), rotation = obj.getRotation(), smooth = false}
+                        destroyObject(obj)
+                        data.defendBag.takeObject(attrs)
+                    elseif not fastMode then
+                        data.tints[suffix] = obj.getColorTint()
+                        if not data.objects[suffix] then
+                            data.objects[suffix] = {obj}
+                        else
+                            table.insert(data.objects[suffix], obj)
+                        end
+                    end
+                end
+            end
+        end
+    end
+
+    -- Pass 2: Iterate over found objects and swap color tints and object names.
+    -- After we're done, put objects in their new presence bag, if applicable.
+    if fastMode then
+        -- All's we did is maybe recolor some defence tokens, so we can skip the rest of this.
```

Comment:
if you're handling defence tokens while the color is unpicked, you should also handle isolate tokens as well in the same manner

Reply:
Fixed now.  Isolate token bags didn't exist before, so the same trickery that redoes presence colors was sufficient then.

Reply:
could you update this comment then?